stetho/src/main/java/com/facebook/stetho/dumpapp/plugins/HprofDumperPlugin.java

```
@@ -0,0 +1,121 @@
+ * <p/>
+ * For convenient usage, see: {@code scripts/hprof_dump.sh}
+ */
+public class HprofDumperPlugin implements DumperPlugin {
+
+  private static final String NAME = "hprof";
+  private final Context mContext;
+
+  public HprofDumperPlugin(Context context) {
+    mContext = context;
+  }
+
+  @Override
+  public String getName() {
+    return NAME;
+  }
+
+  @Override
+  public void dump(DumperContext dumpContext) throws DumpException {
+    final PrintStream output = dumpContext.getStdout();
+
+    Iterator<String> argsIter = dumpContext.getArgsAsList().iterator();
+    String outputPath = argsIter.hasNext() ? argsIter.next() : null;
+    if (outputPath == null) {
+      usage(output);
+    } else {
+      if ("-".equals(outputPath)) {
+        handlePipeOutput(output);
+      } else {
+        File outputFile = new File(outputPath);
+        if (!outputFile.isAbsolute()) {
+          outputFile = mContext.getFileStreamPath(outputPath);
+        }
+        writeHprof(outputFile);
+        output.println("Wrote to " + outputFile);
+      }
+    }
+  }
+
+  private void handlePipeOutput(OutputStream output) throws DumpException {
+    File hprofFile = mContext.getFileStreamPath("hprof-dump.hprof");
+    try {
+      writeHprof(hprofFile);
+      try {
+        InputStream input = new FileInputStream(hprofFile);
+        try {
+          Util.copy(input, output, new byte[2048]);
+        } finally {
+          input.close();
+        }
+      } catch (IOException e) {
+        throw new DumpException("Failure copying " + hprofFile + " to dumper output");
+      }
+    } finally {
+      if (hprofFile.exists()) {
+        hprofFile.delete();
+      }
+    }
+  }
+
+  private void writeHprof(File outputPath) throws DumpException {
+    try {
+      // Test that we can write here.  dumpHprofData appears to hang if it cannot write
+      // to the target location on ART.
+      truncateAndDeleteFile(outputPath);
```

This is concerning manual usage as per:
```shell
dumpapp hprof /sdcard/foo
```

If you can't write to `/sdcard/foo`, it silently fails due to a bug in ART.  This code tests for that possibility before it calls into the system.